File: src/find-players-with-zero-or-one-losses/find-players-with-zero-or-one-losses.py
# https://leetcode.com/problems/find-players-with-zero-or-one-losses

class Solution(object):
    def init_score(self, score, players):
        for p in players:
            if p not in score:
                score[p] = [0, 0]

    def findWinners(self, matches):
        """
        :type matches: List[List[int]]
        :rtype: List[List[int]]
        """
        score = dict()
        for (w, l) in matches:
            self.init_score(score, [w, l])
            score[w][0] += 1
            score[l][1] += 1
        a, b = [], []
        for key, (w, l) in score.items():
            if l == 0:
                a.append(key)
            elif l == 1:
                b.append(key)
        a.sort()
        b.sort()
        return [a, b]

if __name__ == "__main__":
    sol = Solution()
    tests = [
        (
            [[1,3],[2,3],[3,6],[5,6],[5,7],[4,5],[4,8],[4,9],[10,4],[10,9]],
            [[1,2,10],[4,5,7,8]],
        ),
        (
            [[2,3],[1,3],[5,4],[6,4]],
            [[1,2,5,6],[]],
        ),
    ]
    for (matches, solution) in tests:
        result = sol.findWinners(matches)
        assert result == solution, \
            f"result {result} != solution {solution}"
    print("✅ All tests passed")

File: src/find-players-with-zero-or-one-losses/find-players-with-zero-or-one-losses.rs
// https://leetcode.com/problems/find-players-with-zero-or-one-losses
struct Solution;

use std::collections::HashMap;

impl Solution {
    pub fn find_winners(matches: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut score = HashMap::new();
        let mut a = vec![];
        let mut b = vec![];

        for pair in matches {
            score.entry(pair[0]).or_insert((0, 0)).0 += 1;
            score.entry(pair[1]).or_insert((0, 0)).1 += 1;
        }

        for (id, (_w, l)) in score {
            if l == 0 {
                a.push(id);
            } else if l == 1 {
                b.push(id);
            }
        }

        a.sort();
        b.sort();

        vec![a, b]
    }
}

fn main() {
    let tests = vec![
        (
            vec![
                vec![1, 3],
                vec![2, 3],
                vec![3, 6],
                vec![5, 6],
                vec![5, 7],
                vec![4, 5],
                vec![4, 8],
                vec![4, 9],
                vec![10, 4],
                vec![10, 9],
            ],
            vec![vec![1, 2, 10], vec![4, 5, 7, 8]],
        ),
        (
            vec![vec![2, 3], vec![1, 3], vec![5, 4], vec![6, 4]],
            vec![vec![1, 2, 5, 6], vec![]],
        ),
    ];

    for (matches, solution) in tests {
        let result = Solution::find_winners(matches);
        assert_eq!(result, solution);
    }

    println!("✅ All tests passed")
}
